File: Programmers/[72411] 메뉴 리뉴얼/woong-jae/72411.js
function solution(orders, course) {
    
    const getCombination = arr => {
        const comb = (arr, pick) => {
            if(pick === 1) return arr.map(elem => [elem]);
            const combs = [];
            arr.forEach((elem, index) => {
                const smallerCombs = comb(arr.slice(index + 1), pick - 1);
                smallerCombs.forEach(smallComb => combs.push([elem].concat(smallComb)));
            });
            return combs;
        }
        
        const ret = [];
        for(let len = 2; len <= arr.length; len++) {
            ret.push(...comb(arr, len).map(menu => menu.join("")));
        }
        return ret;
    }
    
    let menu_count = new Map();
    
    orders.forEach(order => {
        const allMenu = getCombination(order.split("").sort());
        allMenu.forEach(menu => {
            if(!menu_count.has(menu)) menu_count.set(menu, 0);
            menu_count.set(menu, menu_count.get(menu) + 1);
        });
    });
    
    menu_count = Array.from(menu_count);
    
    return course.map(courseLen => {
        const cand = menu_count.filter(([menu, count]) => menu.length === courseLen && count >= 2)
        const max_count = cand.reduce((prev, cur) => Math.max(prev, cur[1]), 0);
        return cand
            .filter(([_, count]) => count === max_count)
            .map(([menu]) => menu);
    }).flat().sort();
}

File: Programmers/[72411] 메뉴 리뉴얼/Johoseong/72411.py
from itertools import combinations
def solution(orders, course):
    answer = []
    combis = dict()
    for i in range(len(course)):
        combis[course[i]] = dict()

    for c in course:
        for o in orders:
            coms = list(combinations(o, c)) # 각 order마다 course개 조합 구함
            for com in coms:
                tmp = ''.join(sorted(com))
                combis[c].setdefault(tmp, 0)
                combis[c][tmp] += 1 # 해당 코스 조합이 나온 횟수 count
    
    for i in combis: # 제일 많이 주문된 코스조합 구하기
        MAX = 2 # 2명 이상 손님이 주문한 경우만 됨
        tmp = []
        for j in combis[i]:
            if combis[i][j] > MAX:
                MAX = combis[i][j]
                tmp = [j]
            elif combis[i][j] == MAX:
                tmp.append(j)
        for t in tmp:
            answer.append(t)

    answer.sort()
    return answer